core/src/storage2/traits/layout/impls.rs

```diff
@@ -0,0 +1,242 @@
+                    StructLayout::new(vec![
+                        $(
+                            FieldLayout::new(None, <$frag as StorageLayout>::layout(key_ptr)),
+                        )*
+                    ])
+                )
+            }
+        }
+    }
+}
+impl_layout_for_tuple!(A);
+impl_layout_for_tuple!(A, B);
+impl_layout_for_tuple!(A, B, C);
+impl_layout_for_tuple!(A, B, C, D);
+impl_layout_for_tuple!(A, B, C, D, E);
+impl_layout_for_tuple!(A, B, C, D, E, F);
+impl_layout_for_tuple!(A, B, C, D, E, F, G);
+impl_layout_for_tuple!(A, B, C, D, E, F, G, H);
+impl_layout_for_tuple!(A, B, C, D, E, F, G, H, I);
+impl_layout_for_tuple!(A, B, C, D, E, F, G, H, I, J);
+
+impl<T> StorageLayout for Box<T>
+where
+    T: StorageLayout,
+{
+    fn layout(key_ptr: &mut KeyPtr) -> Layout {
+        <T as StorageLayout>::layout(key_ptr)
+    }
+}
+
+impl<T> StorageLayout for Option<T>
+where
+    T: StorageLayout,
+{
+    fn layout(key_ptr: &mut KeyPtr) -> Layout {
+        let dispatch_key = key_ptr.advance_by(1);
+        Layout::Enum(EnumLayout::new(
+            dispatch_key,
+            vec![
+                (
+                    Discriminant::from(0),
+                    StructLayout::new(vec![FieldLayout::new(
+                        None,
+                        <T as StorageLayout>::layout(&mut key_ptr.clone()),
+                    )]),
+                ),
+                (Discriminant::from(1), StructLayout::new(Vec::new())),
+            ],
+        ))
+    }
+}
+
+impl<T, E> StorageLayout for Result<T, E>
+where
+    T: StorageLayout,
+    E: StorageLayout,
+{
+    fn layout(key_ptr: &mut KeyPtr) -> Layout {
+        let dispatch_key = key_ptr.advance_by(1);
+        Layout::Enum(EnumLayout::new(
+            dispatch_key,
+            vec![
+                (
+                    Discriminant::from(0),
+                    StructLayout::new(vec![FieldLayout::new(
+                        None,
+                        <T as StorageLayout>::layout(&mut key_ptr.clone()),
+                    )]),
+                ),
+                (
+                    Discriminant::from(1),
+                    StructLayout::new(vec![FieldLayout::new(
+                        None,
+                        <E as StorageLayout>::layout(&mut key_ptr.clone()),
+                    )]),
+                ),
+            ],
+        ))
+    }
+}
+
+impl<T> StorageLayout for Vec<T>
+where
+    T: Metadata + 'static,
+{
+    fn layout(key_ptr: &mut KeyPtr) -> Layout {
+        Layout::Cell(CellLayout::new::<Self>(LayoutKey::from(
+            key_ptr.advance_by(1),
+        )))
+    }
+}
+
+impl<K, V> StorageLayout for BTreeMap<K, V>
+where
+    K: Metadata + 'static,
+    V: Metadata + 'static,
+{
+    fn layout(key_ptr: &mut KeyPtr) -> Layout {
+        Layout::Cell(CellLayout::new::<Self>(LayoutKey::from(
+            key_ptr.advance_by(1),
+        )))
+    }
+}
+
+// impl<T> StorageLayout for BinaryHeap<T>
```

Comment:
These are commented out because they ideally should exist but are currently somehow not supported by the `type_metadata` crate which they should.